Add a tournament management feature

Even593 opened the pull request:
**Swiss tournament support for admin panel**

- add tournaments, tournament_rounds, tournament_matches, and tournament_standings tables for recording match results and score
- implement Swiss pairing, match execution, and standings helpers directly in app.py
- added TournamentManager component to admin UI

Even593 committed:
Add a tournament management feature, including APIs and front-end interfaces for creating, viewing, and updating tournaments.

Wxvxl reviewed (changes requested):
Lots of changes needed in this. Feels like need some heavy refactoring as lots of the code feels inefficient, unnecessary and is really hard to read. On top of that, one of the main issue is that the games are still hardcoded to be played with two agents which is something that needs to be refactored and is already an issue.

Even593 committed:
Change the competition score type from Decimal to Integer, update the database table structure to reflect this change, and adjust the frontend display logic accordingly.

Even593 commented on PythonWebserver/app.py:
```diff
@@ -94,6 +96,283 @@ def fetch_latest_agent(groupname, game):
 
+# ------ Tournament Functions Below ------ #    
+
+def fetch_latest_agents_for_game(cur, game):
```
it fetch every group's latest agents, when admin star a new tournament.
If it reuse `fetch_latest_agent` fetches only one group’s most recent agent and manages its own DB connection, so reusing it would loop through every group and repeatedly opening connections.

Even593 commented on PythonWebserver/app.py:
```diff
@@ -94,6 +96,283 @@ def fetch_latest_agent(groupname, game):
+        winner_label = agent2_info["agent_name"]
+    else:
+        winner_agent_id = None
+        agent1_score = Decimal("0.5")
```
Honestly, the the score rules comes from wikipedia of `Swiss-system tournament
`. I'd like to use an easier way to calculate the score if you agree.

Even593 commented:
Some of the logic is confusing and contains unnecessary attributes, it needs to be refactored. but we need to clarify the tournament mechanism before refactor.

Even593 committed:
Merge remote-tracking branch 'origin' into feature/tournament

Even593 committed:
feat: Update tournament structure and logic for knockout format

- modified TournamentManager component to reflect knockout tournament
- enhanced agent match logic to handle results and decisions more robustly.
- adopt new format of result of game.play

Even593 committed:
feat: Update tournament management

Even593 committed:
Improve function docstrings for clarity and consistency

Even593 committed:
Row_winner descr

Even593 committed:
remove unused import

Even593 commented:
**Tournament feature update**
- Switch tournament creation to true single-elimination: random shuffle, per-round bracket, byes + tiebreak handling, and points logic, which makes match logic more straightforward and less attributes. 
- Capture per-match metadata (raw winner/tiebreak/advancing agent) for client further check
- Update admin UI (TournamentManager) to match knockout mode

unicorn-seer commented on competitive-agent-site/src/components/admin/TournamentManager.tsx:
```diff
@@ -0,0 +1,317 @@
+
+const GAME_OPTIONS = [
+  { value: "conn4", label: "Conn 4" },
+  { value: "tictactoe", label: "Tic Tac Toe" },
```
New game added for rock paper scissors in pr #44, can you please review the pr and incorporate it into the tournament options if it's possible, thanks.

Even593 replied:
Will check it by today.

unicorn-seer commented on PythonWebserver/dbSetup.py:
```diff
@@ -134,8 +138,62 @@
+    rounds INT,
+    status VARCHAR(20) NOT NULL DEFAULT 'pending',
+    created_by INT REFERENCES users(user_id),
+    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
```
Currently the timestamp does not reflect perth timezone, setup configuration has already been updated in compose file through pr #44, can you please change the code when using TIMESTAMP to:
 
```suggestion
    created_at TIMESTAMPTZ DEFAULT CURRENT_TIMESTAMP
```

Even593 replied:
good point, thanks for suggestion.

unicorn-seer reviewed (commented):
Locally tested, the feature works fine, a few comments left for integration.

Wxvxl reviewed (approved):
Tested Locally, still need some adjustments to make the code cleaner but should be working fine but is merge ready for now.
